src/_pytest/config/__init__.py

```diff
@@ -392,40 +393,48 @@ def parse_hookimpl_opts(self, plugin: _PluggyPlugin, name: str):
         if name == "pytest_plugins":
             return
 
-        method = getattr(plugin, name)
         opts = super().parse_hookimpl_opts(plugin, name)
+        if opts is not None:
+            return opts
 
+        method = getattr(plugin, name)
         # Consider only actual functions for hooks (#3775).
         if not inspect.isroutine(method):
             return
-
         # Collect unmarked hooks as long as they have the `pytest_' prefix.
-        if opts is None and name.startswith("pytest_"):
-            opts = {}
-        if opts is not None:
-            # TODO: DeprecationWarning, people should use hookimpl
-            # https://github.com/pytest-dev/pytest/issues/4562
-            known_marks = {m.name for m in getattr(method, "pytestmark", [])}
 
-            for name in ("tryfirst", "trylast", "optionalhook", "hookwrapper"):
-                opts.setdefault(name, hasattr(method, name) or name in known_marks)
+        known_marks = {m.name for m in getattr(method, "pytestmark", [])}
+        opts = {
+            name: hasattr(method, name) or name in known_marks
+            for name in ("tryfirst", "trylast", "optionalhook", "hookwrapper")
+        }
+        if any(opts.values()):
+            message = _pytest.deprecated.HOOK_LEGACY_MARKING.format(
+                type="spec", fullname=method.__qualname__
+            )
+            warn_explicit_for(method, message)
+
         return opts
 
     def parse_hookspec_opts(self, module_or_class, name: str):
         opts = super().parse_hookspec_opts(module_or_class, name)
         if opts is None:
             method = getattr(module_or_class, name)
-
             if name.startswith("pytest_"):
-                # todo: deprecate hookspec hacks
-                # https://github.com/pytest-dev/pytest/issues/4562
+
                 known_marks = {m.name for m in getattr(method, "pytestmark", [])}
                 opts = {
                     "firstresult": hasattr(method, "firstresult")
                     or "firstresult" in known_marks,
                     "historic": hasattr(method, "historic")
                     or "historic" in known_marks,
                 }
+                # hook from xdist, fixing in  ...
```

Comment:
Interesting, went ahead and fixed this in pytest-xdist: https://github.com/pytest-dev/pytest-xdist/pull/708

Perhaps with the fix we can drop this? I think it is reasonable to ask users to update pytest-xdist if they want to get rid of this warning.